components/script/dom/htmltrackelement.rs

```diff
@@ -2,42 +2,124 @@
 use dom_struct::dom_struct;
 use html5ever::{LocalName, Prefix};
 
 #[dom_struct]
 pub struct HTMLTrackElement {
     htmlelement: HTMLElement,
+    ready_state: u16,
+    track: Dom<TextTrack>,
 }
 
 impl HTMLTrackElement {
     fn new_inherited(
         local_name: LocalName,
         prefix: Option<Prefix>,
         document: &Document,
+        track: &TextTrack,
     ) -> HTMLTrackElement {
         HTMLTrackElement {
             htmlelement: HTMLElement::new_inherited(local_name, prefix, document),
+            ready_state: HTMLTrackElementConstants::NONE,
+            track: Dom::from_ref(&track),
         }
     }
 
-    #[allow(unrooted_must_root)]
     pub fn new(
         local_name: LocalName,
         prefix: Option<Prefix>,
         document: &Document,
     ) -> DomRoot<HTMLTrackElement> {
+        let track = TextTrack::new(
+            &document.window(),
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            Default::default(),
+            Default::default(),
+        );
         Node::reflect_node(
             Box::new(HTMLTrackElement::new_inherited(
-                local_name, prefix, document,
+                local_name, prefix, document, &track,
             )),
             document,
             HTMLTrackElementBinding::Wrap,
         )
     }
 }
+
+impl HTMLTrackElementMethods for HTMLTrackElement {
+    // https://html.spec.whatwg.org/multipage/#dom-track-kind
+    fn Kind(&self) -> DOMString {
+        let element = self.upcast::<Element>();
+        // The the value of "kind" and transform all uppercase
+        // chars into lowercase.
+        let kind = element
+            .get_string_attribute(&local_name!("kind"))
+            .to_lowercase();
+        match &*kind {
+            "subtitles" | "captions" | "descriptions" | "chapters" | "metadata" => {
+                // The value of "kind" is valid. Return the lowercase version
+                // of it.
+                DOMString::from(kind)
+            },
+            _ if kind.is_empty() => {
+                // The default value should be "subtitles". If "kind" has not
+                // been set, the real value for "kind" is "subtitles"
+                DOMString::from("subtitles")
+            },
+            _ => {
+                // If "kind" has been set but it is not one of the valid
+                // values, return the default invalid value of "metadata"
+                DOMString::from("metadata")
+            },
+        }
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-track-kind
+    // Do no transformations on the value of "kind" when setting it.
+    // All transformations should be done in the get method.
+    make_setter!(SetKind, "kind");
+
+    // https://html.spec.whatwg.org/multipage/#dom-track-src
+    make_url_getter!(Src, "src");
+    // https://html.spec.whatwg.org/multipage/#dom-track-src
+    make_url_setter!(SetSrc, "src");
```

Comment:
I believe we depend on implementing and using `Element::set_url_attribute` within `make_url_setter` to be fully spec compliant here. Could you file a follow up bug to implement this, please?

Reply:
Yeah, for now `make_url_setter` is the same as `make_setter`.